File: overlay/src/engine_policy.cpp
#include "engine_policy.h"

#include <algorithm>
#include <cctype>

namespace overlay {

std::vector<std::string> EnginePolicy::selectDefaults(const std::string& engine) {
    // 安全档（默认开）：基础清理 + 常用去重（带长度/非重复保护）。
    // 激进过滤器（incremental_dedup / dedup_mixed_lines / shift_jis / quote_only /
    // line_trimmer 等）默认关闭，需用户在详情页手动开启——避免误伤正常字幕
    // （叠词、短重复、ABAB 等），与 LunaTranslator 默认保守一致。
    const std::vector<std::string> kSafe = {
        "furigana", "control_char", "dedup_chars", "dedup_lines", "unicode_normalize"};

    // 深度 Hook 引擎：输出干净，最小清洗
    if (isDeepHookEngine(engine)) {
        return {"control_char", "unicode_normalize"};
    }
    // krkr 系：注音 + 重复字符 + 整句重复（安全档）
    if (containsAny(engine, {"krkr", "KiriKiri", "krkr2", "krkrZ"})) {
        return kSafe;
    }
    // Unity：安全档
    if (containsAny(engine, {"Unity"})) {
        return kSafe;
    }
    // RPGMaker：安全档（去掉了 line_trimmer 激进项）
    if (containsAny(engine, {"RPGMaker", "RPG Maker"})) {
        return kSafe;
    }
    // Light.vn：重复字符 + 控制字符 + 正规化
    if (containsAny(engine, {"Light.vn", "light.vn", "Lightvn"})) {
        return {"dedup_chars", "control_char", "unicode_normalize"};
    }
    // TyranoScript：HTML 标签 + 安全档
    if (containsAny(engine, {"TyranoScript", "Tyrano"})) {
        return {"html_tag", "furigana", "control_char", "unicode_normalize"};
    }
    // 未知引擎：安全档（不再全开 shift_jis 等激进项）
    return kSafe;
}

bool EnginePolicy::isDeepHookEngine(const std::string& engine) {
    return containsAny(engine, {"Ren'Py", "Renpy", "renpy"});
}

bool EnginePolicy::containsAny(const std::string& engine,
                               const std::vector<std::string>& keywords) {
    std::string lower;
    lower.reserve(engine.size());
    for (char c : engine) {
        lower += static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
    }
    for (const auto& kw : keywords) {
        std::string kl;
        kl.reserve(kw.size());
        for (char c : kw) {
            kl += static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
        }
        if (lower.find(kl) != std::string::npos) return true;
    }
    return false;
}

}  // namespace overlay

File: kazari_play/utils/engine_policy.py
"""引擎→清洗过滤器默认策略（镜像 overlay/src/engine_policy.cpp，须保持一致）

未启动游戏时，前端清洗配置面板据此显示"引擎默认勾选"的过滤器。
游戏运行中则由 C++ 返回真实生效配置。
"""

# 深度 Hook 引擎关键词（输出完整句，几乎无污染）
_DEEP_HOOK = ("ren'py", "renpy")

# 安全档（默认开）：基础清理 + 常用去重（带长度/非重复保护）。
# 激进过滤器（incremental_dedup / dedup_mixed_lines / shift_jis / quote_only /
# line_trimmer 等）默认关闭，需用户在详情页手动开启——避免误伤正常字幕
# （叠词、短重复、ABAB 等），与 LunaTranslator 默认保守一致。
_SAFE = ["furigana", "control_char", "dedup_chars", "dedup_lines", "unicode_normalize"]

# 引擎关键词 → 默认启用的过滤器（按顺序，order 依次递增）
_ENGINE_RULES = [
    (("krkr", "kirikiri", "krkr2", "krkrz"), list(_SAFE)),
    (("unity",), list(_SAFE)),
    (("rpgmaker", "rpg maker"), list(_SAFE)),
    (("light.vn", "lightvn"), ["dedup_chars", "control_char", "unicode_normalize"]),
    (("tyranoscript", "tyrano"), ["html_tag", "furigana", "control_char", "unicode_normalize"]),
]


def select_defaults(engine: str) -> list:
    """返回引擎默认启用的过滤器 ID 列表（顺序即执行顺序）"""
    e = (engine or "").lower()
    for kw in _DEEP_HOOK:
        if kw in e:
            return ["control_char", "unicode_normalize"]
    for keywords, filters in _ENGINE_RULES:
        for kw in keywords:
            if kw in e:
                return list(filters)
    return list(_SAFE)


def default_filter_config(engine: str) -> list:
    """返回带 enabled/order 的过滤器配置（供前端显示引擎默认勾选）"""
    return [{"id": fid, "enabled": True, "order": i}
            for i, fid in enumerate(select_defaults(engine))]
